chips/sam4l/src/eic.rs

```diff
@@ -0,0 +1,339 @@
+//! Implementation of the SAM4L External Interrupt Controller (EIC).
+//!
+//! Datasheet section "21. External Interrupt Controller (EIC)".
+//!
+//! The External Interrupt Controller (EIC) allows pins to be configured as external
+//! interrupts. Each external interrupt has its own interrupt request and can be individually
+//! interrupted. Each external interrupt can generate an interrupt on rising or falling edge, or
+//! high or low level. Every interrupt input has a configurable filter to remove spikes from
+//! the interrupt source. Every interrupt pin can also be configured to be asynchronous in order
+//! to wake up the part from sleep modes where the CLK_SYNC clock has been disabled.
+//!
+// Author: Josh Zhang <jiashuoz@cs.princeton.edu>
+// Last modified July 9, 2019
+
+use crate::pm::{self, Clock, PBDClock};
+use core::cell::Cell;
+use kernel::common::cells::OptionalCell;
+use kernel::common::registers::{register_bitfields, ReadOnly, ReadWrite, WriteOnly};
+use kernel::common::StaticRef;
+use kernel::hil;
+
+/// Enum for enabling or disabling spurious event filtering (i.e. de-bouncing control).
+pub enum FilterMode {
+    FilterEnable,
+    FilterDisable,
+}
+
+/// Enum for selecting synchronous or asynchronous mode. Interrupts in asynchronous mode
+/// can wake up the system from deep sleep mode.
+pub enum SynchronizationMode {
+    Synchronous,
+    Asynchronous,
+}
+
+/// The sam4l chip supports 9 external interrupt lines: Ext1 - Ext8 and an additional
+/// Non-Maskable Interrupt (NMI) pin. NMI has the same properties as the other external
+/// interrupts, but is connected to the NMI request of the CPU, enabling it to interrupt
+/// any other interrupt mode.
+#[derive(Copy, Clone, Debug)]
+#[repr(u32)]
+pub enum Line {
+    Nmi = 1,
+    Ext1 = 2,
+    Ext2 = 4,
+    Ext3 = 8,
+    Ext4 = 16,
+    Ext5 = 32,
+    Ext6 = 64,
+    Ext7 = 128,
+    Ext8 = 256,
+}
+
+#[repr(C)]
+struct EicRegisters {
+    ier: WriteOnly<u32, Interrupt::Register>,          // 0x00
+    idr: WriteOnly<u32, Interrupt::Register>,          // 0x04
+    imr: ReadOnly<u32, Interrupt::Register>,           // 0x08
+    isr: ReadOnly<u32, Interrupt::Register>,           // 0x0c
+    icr: WriteOnly<u32, Interrupt::Register>,          // 0x10
+    mode: ReadWrite<u32, Interrupt::Register>,         // 0x14
+    edge: ReadWrite<u32, Interrupt::Register>,         // 0x18
+    level: ReadWrite<u32, Interrupt::Register>,        // 0x1c
+    filter: ReadWrite<u32, Interrupt::Register>,       // 0x20
+    test: ReadWrite<u32, Test::Register>,              // 0x24
+    asynchronous: ReadWrite<u32, Interrupt::Register>, // 0x28
+    _reserved0: ReadOnly<u32>,                         // 0x02c, skip
+    en: WriteOnly<u32, Interrupt::Register>,           // 0x30
+    dis: WriteOnly<u32, Interrupt::Register>,          // 0x34
+    ctrl: ReadOnly<u32, Interrupt::Register>,          // 0x38
+}
+
+// IER: Writing a one to this bit will set the corresponding bit in IMR.
```

Comment:
Can you move all these comments into doc comments above their register (like https://github.com/tock/tock/blob/master/chips/nrf52/src/adc.rs#L10)?

It's a bit of a newer Tock convention, but a good one I think